onnx/defs/schema.h

```diff
@@ -27,15 +27,20 @@ namespace ONNX_NAMESPACE {
 
 struct FunctionBodyBuildContext {
   virtual const AttributeProto* getAttribute(const std::string& name) const = 0;
-  virtual bool hasInput(int i) const = 0;
-  virtual bool hasOutput(int i) const = 0;
+  virtual bool hasInput(int inputIndex) const = 0;
+  virtual bool hasOutput(int inputIndex) const = 0;
+  // getInputType(i) should return null for missing optional inputs, or if
+  // type-inference could not infer the input-type (erroneous model).
+  virtual const TypeProto* getInputType(int inputIndex) const = 0;
```

Comment:
Maybe we should throw if a type is missing? We can write another helper function `virtual bool hasInputType(int inputIndex)` for checking if input type exists before calling `getInputType(int inputIndex)`.

Reply:
I think this is simpler than throwing an exception. We already have optional inputs to handle anyway. It is good to make the caller be aware and handle the possibility explicitly. (It is easier to ignore/forget to handle exceptions.)